src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,830 @@
+        {
+            dims = compute_dims(op, dims);
+            if(not apply_reshape(dims))
+                return false;
+        }
+        else if(op.name() == "transpose")
+        {
+            dims = compute_dims(op, dims);
+            if(not apply_transpose(v["permutation"].to_vector<std::int64_t>()))
+                return false;
+        }
+        else if(op.name() == "multibroadcast")
+        {
+            dims = compute_dims(op, dims);
+            // cppcheck-suppress knownConditionTrueFalse
+            if(not apply_broadcast(dims))
+                return false;
+        }
+        else if(op.name() == "broadcast")
+        {
+            dims = compute_dims(op, dims);
+            // cppcheck-suppress knownConditionTrueFalse
+            if(not apply_broadcast(dims, v["axis"].to<std::size_t>()))
+                return false;
+        }
+        else if(op.name() != "contiguous")
+        {
+            return false;
+        }
+    }
+    return true;
+}
+bool shape_transform_descriptor::apply_reshape(const std::vector<std::size_t>& rdims)
+{
+    assert(migraphx::elements(rdims) == this->elements());
+    std::vector<dimension> new_dims;
+    auto subs     = get_all_subdimensions(dimensions);
+    std::size_t i = 0;
+    std::size_t r = 0;
+    while(i < subs.size() and r < rdims.size())
+    {
+        const auto& sub = subs[i];
+        auto idim       = sub.len;
+        auto rdim       = rdims[r];
+        if(idim == rdim)
+        {
+            new_dims.push_back({{sub}});
+        }
+        // squeeze
+        else if(rdim > idim)
+        {
+            auto start = subs.begin() + i;
+            auto it = compute_end_dim(start, subs.end(), rdim, std::mem_fn(&dimension::sub::len));
+            if(it == start)
+                return false;
+            auto n = it - start;
+            i += n;
+            new_dims.push_back({{start, it + 1}});
+        }
+        // unsqueeze
+        else // if(rdim < idim)
+        {
+            auto start = rdims.begin() + r;
+            auto it    = compute_end_dim(start, rdims.end(), idim, id{});
+            if(it == start)
+                return false;
+            auto n = it - start;
+            r += n;
+            transform(range(n + 1), std::back_inserter(new_dims), [&](auto j) -> dimension {
+                auto new_sub = sub;
+                if(not new_sub.axis.empty())
+                    new_sub.axis.push_back(j);
+                new_sub.len = start[j];
+                return {{new_sub}};
+            });
+        }
+        r++;
+        i++;
+    }
+
+    // Handle trailing 1s
+    if(new_dims.size() < rdims.size() and not new_dims.empty())
+    {
+        for(auto d : range(rdims.begin() + new_dims.size(), rdims.end()))
+        {
+            if(d != 1)
+                return false;
+            new_dims.push_back({{dimension::sub{1}}});
+        }
+    }
+
+    if(rdims.size() != new_dims.size())
```

Comment:
assert would suffice here since for loop earlier would make sure both are of same size.

Reply:
I can add an assert, but I still would like this branch so if something goes wrong in a release version it will just bail.